drivers/network/meson/ethernet.c

```diff
@@ -246,13 +250,23 @@ static void eth_setup(void)
     eth_mac->macaddr0hi = h;
 
     eth_dma->busmode = PRIORXTX_11 | ((DMA_PBL << TX_PBL_SHFT) & TX_PBL_MASK);
-    eth_dma->opmode = STOREFORWARD;
+    /*
+     * Operate in store-and-forward mode.
+     * Send pause frames when there's only 1k of fifo left,
+     * stop sending them when there is 2k of fifo left.
+     * Continue DMA on 2nd frame while updating status on first
+     */
+    eth_dma->opmode = STOREFORWARD | EN_FLOWCTL | (0 << FLOWCTL_SHFT) | (0 < DISFLOWCTL_SHFT) | TX_OPSCND;
```

Comment:
Comment doesn't  match code.  Ideally it should be 1<< DISFLOWCTL_SHIFT to get the hysteresis (2k stop point)
From the Linux driver (RFA is the FLOWCTL_SHIFT; RFD DISFLOWCTL_SHIFT fields)

```
/* RFD and RFA fields are encoded as follows
 *
 *   Bit Field
 *   0,00 - Full minus 1KB (only valid when rxfifo >= 4KB and EFC enabled)
 *   0,01 - Full minus 2KB (only valid when rxfifo >= 4KB and EFC enabled)
 *   0,10 - Full minus 3KB (only valid when rxfifo >= 4KB and EFC enabled)
 *   0,11 - Full minus 4KB (only valid when rxfifo > 4KB and EFC enabled)
 *   1,00 - Full minus 5KB (only valid when rxfifo > 8KB and EFC enabled)
 *   1,01 - Full minus 6KB (only valid when rxfifo > 8KB and EFC enabled)
 *   1,10 - Full minus 7KB (only valid when rxfifo > 8KB and EFC enabled)
 *   1,11 - Reserved
 *
 * RFD should always be > RFA for a given FIFO size. RFD == RFA may work,
 * but packet throughput performance may not be as expected.
 *
 * Be sure that bit 3 in GMAC Register 6 is set for Unicast Pause frame
 * detection (IEEE Specification Requirement, Annex 31B, 31B.1, Pause
 * Description).
 *
 * Be sure that DZPA (bit 7 in Flow Control Register, GMAC Register 6),
 * is set to 0. This allows pause frames with a quanta of 0 to be sent
 * as an XOFF message to the link peer.
 */
 ```

Reply:
Yes, you're right, I missed that I fiddled around with different values when testing. I'll change the disable flow control value to Full minus 2K. Out of curiosity though I tested a few different configurations of these values and saw very little change in throughput from 756Mb/s and cpu utilisation ~ 80%. (Tried both (1K,1K), (1K,2K - what it is now), and (2K,3K).